src/wrappers.jl

```diff
@@ -0,0 +1,442 @@
+
+Base.parent(geom::WrapperGeometry) = geom.geom
+
+function Base.:(==)(g1::WrapperGeometry, g2::WrapperGeometry)
+    all(((a, b),) -> a == b, zip(getgeom(g1), getgeom(g2)))
+end
+
+geointerface_geomtype(trait) = throw(ArgumentError("trait $trait not yet implemented as a GeoInterface geometry wrapper"))
+
+# Interface methods
+# With indexing
+function getgeom(trait::AbstractGeometryTrait, geom::WrapperGeometry{<:Any,<:Any,T}, i) where T
+    isgeometry(T) ? getgeom(trait, parent(geom), i) : parent(geom)[i]
+end
+function getgeom(trait::AbstractGeometryTrait, geom::WrapperGeometry{<:Any,<:Any,T}) where T
+    isgeometry(T) ? getgeom(trait, parent(geom)) : parent(geom)
+end
+
+extent(::AbstractGeometryTrait, geom::WrapperGeometry) =
+    isnothing(geom.extent) && isgeometry(parent(geom)) ? extent(parent(geom)) : geom.extent
+function ngeom(trait::AbstractGeometryTrait, geom::WrapperGeometry{<:Any,<:Any,T}) where T
+    isgeometry(T) ? ngeom(parent(geom)) : length(parent(geom))
+end
+
+for (geomtype, trait, childtype, child_trait, length_check, nesting) in (
+        (:Line, :LineTrait, :Point, :PointTrait, ==(2), 1),
+        (:LineString, :LineStringTrait, :Point, :PointTrait, >=(2), 1),
+        (:LinearRing, :LinearRingTrait, :Point, :PointTrait, >=(3), 1),
+        (:MultiPoint, :MultiPointTrait, :Point, :PointTrait, nothing, 1),
+        (:Polygon, :PolygonTrait, :LinearRing, :LinearRingTrait, nothing, 2),
+        (:MultiLineString, :MultiLineStringTrait, :LineString, :LineStringTrait, nothing, 2),
+        (:MultiCurve, :MultiCurveTrait, :LineString, :AbstractCurveTrait, nothing, 2),
+        (:MultiPolygon, :MultiPolygonTrait, :Polygon, :PolygonTrait, nothing, 3),
+        (:GeometryCollection, :GeometryCollectionTrait, :AbstractGeometry, :AbstractGeometryTrait, nothing, nothing),
+        (:PolyhedralSurface, :PolyhedralSurfaceTrait, :Polygon, :PolygonTrait, nothing, 3),
+        # (:Triangle, :TriangleTrait, :LinearRingTrait, :LinearRing, ==(3), 2),
+        # (:Quad, :QuadTrait, :LinearRingTrait, :LinearRing, ==(4), 2),
+        # (:Pentagon, :PentagonTrait, :LinearRingTrait, :LinearRing, ==(5), 2),
+        # (:Hexagon, :HexagonTrait, :LinearRingTrait, :LinearRing, ==(6), 2),
+        # (:Rectangle, :RectangleTrait, :Point, :PointTrait, nothing, 1), ?
+        # (:TIN, :TINTrait, :Triangle, :TriangleTrait, nothing, 3),
+    )
+
+    # Prepare docstring example
+    childname = lowercase(string(childtype))
+    example_child_geoms = if geomtype == :Polygon
+        "interior, hole1, hole2"
+    else
+        join((string(childname, n) for n in 1:(isnothing(length_check) ? 3 : length_check.x)), ", ")
+    end
+    docstring = """
+        $geomtype
+
+        $geomtype(geom, [extent])
+        $geomtype{Z,M}(geom, [extent])
+
+    ## Arguments
+
+    - `geom`: any object returning $trait from `GeoInterface.trait`, or a vector
+        of objects returning $child_trait.
+    - `extent`: an `Extents.Extent`, or `nothing`.
```

Comment:
What about crs?

Reply:
I was wondering that too. Should geometries have crs attached? Is there a use case? I have that just for `Feature` and `FeatureCollection` now.

For extents I think there is a use case - shapefile geometries all have a precalculated extent but others formats don't and have no way to add them. These wrappers would provide a way to do that without allocating new objects.

Reply:
It's in the interface to request a crs for a geometry, so it would make sense to include here.

Reply:
Ok, will do. Do you think even for points?

Reply:
If not we can merge this

Reply:
Sorry, it's hard to distinguish between commit notifications and comments here. Feel free to ping me for bumps. 

I'd say also add crs for Points. If for any geometry, also for Point.